optimistix/_solver/quasi_newton.py

```diff
@@ -115,12 +78,615 @@ def __call__(
+        `new_vals`. If `index_start >= array.shape[0]`, the matrix is
+        rolled to discard the oldest entry.
+
+    **Returns**:
+
+    An updated `(n, n)` JAX array with `new_vals` inserted, either
+    in-place or after rolling.
+    """
+
+    def _roll_and_set(x, new_vals):
+        # roll matrix
+        x = jnp.empty_like(x).at[:-1, :-1].set(x[1:, 1:])
+
+        # sort history buffer
+        last_index = index_start % array.shape[0]
+        sort_idx = (jnp.arange(array.shape[0]) + last_index + 1) % array.shape[0]
+        new_vals = new_vals[sort_idx]
+
+        # set last row and col
+        x = x.at[-1, :-1].set(new_vals[:-1])
+        return x.at[:, -1].set(new_vals)
+
+    def _set_at_index(x, new_vals):
+        x = x.at[index_start, :-1].set(new_vals[:-1])
+        return x.at[:, index_start].set(new_vals)
+
+    return jax.lax.cond(
+        index_start < array.shape[0], _set_at_index, _roll_and_set, array, new_vals
+    )
+
+
+def _lbfgs_inverse_hessian_operator_fn(
+    pytree: Y,  # TODO name this better?
+    state: _LBFGSInverseHessianUpdateState,
+) -> Y:
+    """
+    LBFGS descent linear operator.
+
+    """
+    history_len = state.inner_history.shape[0]
+    circ_index = (jnp.arange(history_len) + state.index_start) % history_len
+
+    # First loop: iterate backwards and compute alpha coefficients
+    def backward_iter(descent_direction, indx):
+        y_diff, grad_diff, inner = jtu.tree_map(
+            lambda x: x[indx],
+            (state.y_diff_history, state.grad_diff_history, state.inner_history),
+        )
+        alpha = inner * tree_dot(y_diff, descent_direction)
+        descent_direction = (descent_direction**ω - alpha * grad_diff**ω).ω
+        return descent_direction, alpha
+
+    # Second loop: iterate forwards and apply correction using stored alpha
+    def forward_iter(args, indx):
+        descent_direction, alpha = args
+        current_alpha = alpha[indx]
+        inner = state.inner_history[circ_index[indx]]
+        y_diff, grad_diff = jtu.tree_map(
+            lambda x: x[circ_index[indx]],
+            (state.y_diff_history, state.grad_diff_history),
+        )
+        current_beta = inner * tree_dot(grad_diff, descent_direction)
+        descent_direction = (
+            descent_direction**ω + (y_diff**ω * (current_alpha - current_beta))
+        ).ω
+        return (descent_direction, alpha), None
+
+    descent_direction, alpha = jax.lax.scan(
+        backward_iter, pytree, circ_index, reverse=True
+    )
+    latest_y_diff, latest_grad_diff = jtu.tree_map(
+        lambda x: x[(state.index_start - 1) % history_len],
+        (state.y_diff_history, state.grad_diff_history),
+    )
+    y_grad_diff_inner = tree_dot(latest_y_diff, latest_grad_diff)
+    grad_diff_norm_sq = tree_dot(latest_grad_diff, latest_grad_diff)
+    gamma_k = jnp.where(
+        grad_diff_norm_sq > 0, y_grad_diff_inner / grad_diff_norm_sq, 1.0
+    )
```

Comment:
Heads-up that this operation can still be unsafe in some contexts, as you're still actually performing the division by zero. (You just paper over it again afterwards.) This can cause an error in some autodiff or nojit contexts.

The way to avoid this is the 'double where' trick, so that even the thrown-away value is numerically safe:
```python
pred = grad_diff_norm_sq > 0
safe_grad_diff_norm_sq = jnp.where(pred, grad_diff_norm_sq, 1.0)
gamma_k = jnp.where(
    pred, y_grad_diff_inner / safe_grad_diff_norm_sq, 1.0
)
```